### PR TITLE
conferences saving

### PR DESCRIPTION
**Pull request by zugz**
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1156)

### CODE REVIEW

File: toxcore/group.c [hunk 1]
```diff
@@ -880,13 +882,23 @@ static void rejoin_frozen_friend(Group_Chats *g_c, int friendcon_id)
     }
 }
 
+static int g_handle_any_status(void *object, int friendcon_id, uint8_t status, void *userdata)
```
Comment (iphydf) [inline toxcore/group.c]: I think this is already in master, no? Perhaps rebase so it doesn't appear in this PR?

Comment (zugz) [reply toxcore/group.c]: No, I closed that PR. This change is only needed for saving.

**Comment by sudden6** [comment]
Are these CI fails because of this PR or did they exist before?

**Comment by zugz** [comment]
* Tuesday, 2018-10-30 at 11:13 -0700 - sudden6 <notifications@github.com>:

>Are these CI fails because of this PR or did they exist before?

The tsan failure is a long-standing problem. The Cirrus failure was due 
to a complaint about this PR (it didn't like a variable being used only 
for an assertion) which I've now fixed. At the time of writing it's 
failing again, but that's just because the tcp test intermittently fails 
(another long-standing problem).

**Comment by sudden6** [comment]
@iphydf Your lgtm seems to block this PR, what is your current state with this?

File: docs/minpgc.md [hunk 2]
```diff
@@ -102,13 +102,13 @@ actually make more sense in the implementation to have a separate list for
...
-numbers and full peer info for all peers. On reload, all peers would be set as
-frozen.
+Saving is implemented by simply saving all live groups with their group numbers
+and full peer info for all peers. On reload, all peers are set as frozen.
 
-The client would need to support this by understanding that these groups exist
-on start-up (e.g. starting windows for them), and by not automatically killing
-groups on closing the client.
+Clients needs to support this by understanding that groups may exist on
+start-up. Clients should call tox\_conference\_get\_chatlist to obtain them. A
```
Comment (iphydf) [inline docs/minpgc.md]: Use ``` `backticks` ``` for symbol names.

File: toxcore/Messenger.h [hunk 3]
```diff
@@ -796,6 +796,9 @@ uint8_t *messenger_save(const Messenger *m, uint8_t *data);
 
 /* Load a state section.
  *
+ * @param data Data to load
```
Comment (iphydf) [inline toxcore/Messenger.h]: Add "." at the end (or remove it from all the others).

File: toxcore/group.c [hunk 4]
```diff
@@ -1587,6 +1596,22 @@ static int group_name_send(const Group_Chats *g_c, uint32_t groupnumber, const u
...
+ */
+bool group_leave(const Group_Chats *g_c, uint32_t groupnumber)
+{
+    Group_c *g = get_group_c(g_c, groupnumber);
+
+    if (!g) {
+        return false;
+    }
+
+    return (group_kill_peer_send(g_c, groupnumber, g->peer_number) == 0);
```
Comment (iphydf) [inline toxcore/group.c]: Remove extra () around returned expression.

File: toxcore/group.c [hunk 5]
```diff
@@ -2875,22 +2900,242 @@ void send_name_all_groups(Group_Chats *g_c)
...
+    *data = peer->nick_len;
+    ++data;
+
+    memcpy(data, peer->nick, peer->nick_len);
+    data += peer->nick_len;
+
+    return data;
+}
+
+#define SAVED_CONF_SIZE_CONSTANT (1 + GROUP_ID_LENGTH + sizeof(uint32_t) \
```
Comment (iphydf) [inline toxcore/group.c]: This is horrible, but ok.. similar to the rest that is also horrible.

File: toxcore/group.c [hunk 6]
```diff
@@ -2875,22 +2900,242 @@ void send_name_all_groups(Group_Chats *g_c)
...
+        memcpy(g->title, data, g->title_len);
+        data += g->title_len;
+
+        for (uint32_t j = 0; j < g->numfrozen; ++j) {
+            if (length < (uint32_t)(data - init_data) + SAVED_PEER_SIZE_CONSTANT) {
+                return STATE_LOAD_STATUS_ERROR;
+            }
+
+            Group_Peer *peer = &g->frozen[j];
+            memset(peer, 0, sizeof(Group_Peer));
```
Comment (iphydf) [inline toxcore/group.c]: This is not ideal. Are we not setting every field of `Group_Peer` explicitly while loading it?

Comment (zugz) [reply toxcore/group.c]: Well, this is also how the peer structure is initialised in addpeer. Should I add an initialisation function to be used in both places, explicitly setting the various fields to 0?

File: docs/minpgc.md (hunk 2 above)
Comment (zugz) [reply docs/minpgc.md]: done

File: toxcore/Messenger.h (hunk 3 above)
Comment (zugz) [reply toxcore/Messenger.h]: done

**Pull request merged**